File: orro/desk.py
"""High-level desk operations: status, movement, presets, height targeting."""

from __future__ import annotations

import base64
import contextlib
import time
from typing import Any

from orro.cloud import TuyaCloud
from orro.config import Config
from orro.lan import TuyaLan


def movement_commands(
    action: str, config: Config, target: str | None = None
) -> list[dict[str, Any]]:
    """Build the DP command list for a given movement action."""
    wake = [{"code": "child_lock", "value": False}]
    if action == "up":
        return wake + [
            {"code": "move_down", "value": False},
            {"code": "move_up", "value": True},
        ]
    if action == "down":
        return wake + [
            {"code": "move_up", "value": False},
            {"code": "move_down", "value": True},
        ]
    if action == "stop":
        return [{"code": "move_up", "value": False}, {"code": "move_down", "value": False}]
    if action in {"sit", "stand", "goto"}:
        preset = target or config.presets.get(action, "mem1")
        return wake + [{"code": "memory_location", "value": preset}]
    raise ValueError(f"Unknown action: {action}")


def cloud_send(config: Config, commands: list[dict[str, Any]]) -> Any:
    """Send commands via the Cloud API."""
    return TuyaCloud(config).send(commands)


def lan_then_cloud(
    config: Config, commands: list[dict[str, Any]], force_cloud: bool
) -> dict[str, Any]:
    """Try LAN first, fall back to Cloud.  Returns ``{path, result, ...}``."""
    if not force_cloud:
        try:
            result = TuyaLan(config).send_codes(commands)
            return {"path": "lan", "result": result}
        except Exception as exc:
            lan_error = str(exc)
        config.debug(f"lan failed ({lan_error}), falling back to cloud")
        result = cloud_send(config, commands)
        return {"path": "cloud", "lan_error": lan_error, "result": result}
    return {"path": "cloud", "result": cloud_send(config, commands)}


def get_status(config: Config, force_cloud: bool) -> dict[str, Any]:
    """Fetch desk status, preferring LAN when available."""
    if not force_cloud:
        try:
            lan = TuyaLan(config)
            return {
                "path": "lan",
                "ip": lan.ip,
                "version": config.lan_version,
                "status": lan.status(),
            }
        except Exception as exc:
            config.debug(f"lan status failed ({exc}), falling back to cloud")
            cloud_status = TuyaCloud(config).status()
            return {"path": "cloud", "lan_error": str(exc), "status": cloud_status}
    return {"path": "cloud", "status": TuyaCloud(config).status()}


def current_height_mm(status: dict[str, Any], config: Config, path: str) -> int | None:
    """Extract the current height in mm from a status dict."""
    if path == "lan":
        dps = status.get("dps", status)
        value = dps.get(str(config.lan_dps["height_display"])) or dps.get(
            config.lan_dps["height_display"]
        )
    else:
        value = status.get("height_display")
    if value is None:
        return None
    with contextlib.suppress(TypeError, ValueError):
        return int(value)
    return None


def decode_memory_heights(raw_value: str | None) -> dict[str, int]:
    """Decode base64-encoded memory slot heights into ``{memN: mm}``."""
    if not raw_value:
        return {}
    raw = base64.b64decode(raw_value)
    heights: dict[str, int] = {}
    for index in range(0, min(len(raw), 8), 2):
        heights[f"mem{index // 2 + 1}"] = int.from_bytes(raw[index : index + 2], "big")
    return heights


def get_presets(config: Config) -> dict[str, Any]:
    """Fetch preset/memory data from the Cloud shadow properties."""
    properties = TuyaCloud(config).properties()
    memory_heights = decode_memory_heights(properties.get("memory_height", {}).get("value"))
    current_height = properties.get("height_display", {}).get("value")
    memory_location = properties.get("memory_location", {}).get("value")
    return {
        "configured": config.presets,
        "device": memory_heights,
        "device_cm": {key: value / 10 for key, value in memory_heights.items()},
        "current_height_mm": current_height,
        "current_height_cm": current_height / 10 if isinstance(current_height, int) else None,
        "memory_location": memory_location,
        "raw": {
            "memory_height": properties.get("memory_height", {}).get("value"),
            "memory_height_dp": properties.get("memory_height", {}).get("dp_id"),
            "memory_location_dp": properties.get("memory_location", {}).get("dp_id"),
        },
    }


def go_to_height(config: Config, cm: float, force_cloud: bool) -> dict[str, Any]:
    """Move the desk to a specific height in cm, polling until target is reached."""
    target_mm = int(round(cm * 10))
    status_payload = get_status(config, force_cloud)
    path = status_payload["path"]
    status = status_payload["status"]
    current = current_height_mm(status, config, path)
    if current is None:
        raise RuntimeError("height_display is not available; use presets or up/down/stop")
    tolerance_mm = 5
    if abs(current - target_mm) <= tolerance_mm:
        return {
            "path": path,
            "target_mm": target_mm,
            "height_mm": current,
            "result": "already_at_target",
        }
    direction = "up" if target_mm > current else "down"
    commands = movement_commands(direction, config)
    use_cloud = force_cloud or path == "cloud"
    lan_then_cloud(config, commands, use_cloud)
    started = time.time()
    last = current
    try:
        while time.time() - started < 30:
            time.sleep(0.5)
            status_payload = get_status(config, use_cloud)
            last = (
                current_height_mm(status_payload["status"], config, status_payload["path"]) or last
            )
            if abs(last - target_mm) <= tolerance_mm:
                break
            if direction == "up" and last > target_mm:
                break
            if direction == "down" and last < target_mm:
                break
    finally:
        lan_then_cloud(config, movement_commands("stop", config), use_cloud)
    return {"path": path, "target_mm": target_mm, "height_mm": last, "result": "stopped"}

File: internal/desk/desk.go
// Package desk provides high-level operations for a Tuya-based standing desk.
package desk

import (
	"encoding/base64"
	"encoding/binary"
	"fmt"

	"github.com/yashiels/orro-cli/internal/config"
	"github.com/yashiels/orro-cli/internal/tuya"
)

// movementCommands builds the DP command list for a given action.
// target is only used for "goto"; for "sit"/"stand" the preset map is consulted.
func movementCommands(action string, cfg *config.Config, target string) ([]tuya.Command, error) {
	wake := tuya.Command{Code: "child_lock", Value: false}

	switch action {
	case "up":
		return []tuya.Command{
			wake,
			{Code: "move_down", Value: false},
			{Code: "move_up", Value: true},
		}, nil
	case "down":
		return []tuya.Command{
			wake,
			{Code: "move_up", Value: false},
			{Code: "move_down", Value: true},
		}, nil
	case "stop":
		return []tuya.Command{
			{Code: "move_up", Value: false},
			{Code: "move_down", Value: false},
		}, nil
	case "sit", "stand":
		preset, ok := cfg.Presets[action]
		if !ok {
			preset = "mem1"
		}
		return []tuya.Command{wake, {Code: "memory_location", Value: preset}}, nil
	case "goto":
		if target == "" {
			return nil, fmt.Errorf("goto: no target specified")
		}
		return []tuya.Command{wake, {Code: "memory_location", Value: target}}, nil
	default:
		return nil, fmt.Errorf("unknown movement action: %q", action)
	}
}

// MovementCommands is exported for command handlers.
func MovementCommands(action string, cfg *config.Config, target string) ([]tuya.Command, error) {
	return movementCommands(action, cfg, target)
}

// ──────────────────────────────────────────────
// LAN-then-cloud dispatch
// ──────────────────────────────────────────────

// LanThenCloud tries LAN first, falls back to Cloud. Returns {"path","result",...}.
func LanThenCloud(cfg *config.Config, commands []tuya.Command, forceCloud bool) (map[string]any, error) {
	if !forceCloud {
		lan, err := tuya.NewLAN(cfg)
		if err == nil {
			defer lan.Close()
			if sendErr := lan.SendCodes(commands); sendErr == nil {
				return map[string]any{"path": "lan", "result": "ok"}, nil
			} else {
				cfg.Debug(fmt.Sprintf("lan send failed (%v), falling back to cloud", sendErr))
			}
		} else {
			cfg.Debug(fmt.Sprintf("lan unavailable (%v), using cloud", err))
		}
	}

	cloud := tuya.NewCloud(cfg)
	result, err := cloud.Send(commands)
	if err != nil {
		return nil, err
	}
	return map[string]any{"path": "cloud", "result": result}, nil
}

// GetStatus fetches desk status, preferring LAN when available.
func GetStatus(cfg *config.Config, forceCloud bool) (map[string]any, error) {
	if !forceCloud {
		lan, err := tuya.NewLAN(cfg)
		if err == nil {
			defer lan.Close()
			status, statusErr := lan.Status()
			if statusErr == nil {
				return map[string]any{
					"path":    "lan",
					"ip":      lan.IP(),
					"version": lan.Version(),
					"status":  status,
				}, nil
			}
			cfg.Debug(fmt.Sprintf("lan status failed (%v), falling back to cloud", statusErr))
		} else {
			cfg.Debug(fmt.Sprintf("lan unavailable (%v), using cloud", err))
		}
	}

	cloud := tuya.NewCloud(cfg)
	status, err := cloud.Status()
	if err != nil {
		return nil, err
	}
	return map[string]any{"path": "cloud", "status": status}, nil
}

// GetPresets fetches preset/memory data from the Cloud shadow properties.
func GetPresets(cfg *config.Config) (map[string]any, error) {
	cloud := tuya.NewCloud(cfg)
	props, err := cloud.Properties()
	if err != nil {
		return nil, err
	}

	memHeightProp := props["memory_height"]
	var rawMemHeight string
	if memHeightProp != nil {
		rawMemHeight, _ = memHeightProp["value"].(string)
	}
	memHeights := decodeMemoryHeights(rawMemHeight)

	var currentHeight any
	if p := props["height_display"]; p != nil {
		currentHeight = p["value"]
	}

	var currentHeightCM any
	if h, ok := currentHeight.(float64); ok {
		currentHeightCM = h / 10
	}

	var memLoc any
	if p := props["memory_location"]; p != nil {
		memLoc = p["value"]
	}

	deviceCM := make(map[string]float64, len(memHeights))
	for k, v := range memHeights {
		deviceCM[k] = float64(v) / 10
	}

	raw := map[string]any{}
	if memHeightProp != nil {
		raw["memory_height"] = memHeightProp["value"]
		raw["memory_height_dp"] = memHeightProp["dp_id"]
	}
	if p := props["memory_location"]; p != nil {
		raw["memory_location_dp"] = p["dp_id"]
	}

	return map[string]any{
		"configured":        cfg.Presets,
		"device":            memHeights,
		"device_cm":         deviceCM,
		"current_height_mm": currentHeight,
		"current_height_cm": currentHeightCM,
		"memory_location":   memLoc,
		"raw":               raw,
	}, nil
}

// CurrentHeightMM extracts the current height in mm from a status map.
// path is "lan" or "cloud".
func CurrentHeightMM(status map[string]any, cfg *config.Config, path string) (int, bool) {
	var raw any
	if path == "lan" {
		dpNum := cfg.LANDPs["height_display"]
		// LAN status dps are keyed by string DP number or nested under "dps".
		dps, hasDPS := status["dps"].(map[string]any)
		if hasDPS {
			raw = dps[fmt.Sprintf("%d", dpNum)]
		} else {
			raw = status[fmt.Sprintf("%d", dpNum)]
			if raw == nil {
				raw = status["height_display"]
			}
		}
	} else {
		raw = status["height_display"]
	}

	if raw == nil {
		return 0, false
	}

	switch v := raw.(type) {
	case float64:
		return int(v), true
	case int:
		return v, true
	case int64:
		return int(v), true
	default:
		return 0, false
	}
}

// decodeMemoryHeights decodes the base64-encoded memory slot heights.
// Format: 8 bytes, each pair of bytes is one memory slot height in mm (big-endian uint16).
func decodeMemoryHeights(raw string) map[string]int {
	if raw == "" {
		return map[string]int{}
	}
	b, err := base64.StdEncoding.DecodeString(raw)
	if err != nil {
		return map[string]int{}
	}
	heights := make(map[string]int, 4)
	for i := 0; i+1 < len(b) && i < 8; i += 2 {
		slot := i/2 + 1
		heights[fmt.Sprintf("mem%d", slot)] = int(binary.BigEndian.Uint16(b[i : i+2]))
	}
	return heights
}
